dbt_semantic_interfaces/dbt_semantic_interfaces/protocols/dimension.py

```diff
@@ -0,0 +1,41 @@
+from __future__ import annotations
+
+from typing import Optional, Protocol
+
+from dbt_semantic_interfaces.protocols.metadata import Metadata
+from dbt_semantic_interfaces.type_enums.dimension_type import DimensionType
+from dbt_semantic_interfaces.type_enums.time_granularity import TimeGranularity
+
+
+class DimensionValidityParams(Protocol):
+    """Parameters identifying a given dimension as an entity for validity state
+
+    This construct is used for supporting SCD Type II tables, such as might be
+    created via dbt's snapshot feature, or generated via periodic loads from external
+    dimension data sources. In either of those cases, there is typically a time dimension
+    associated with the SCD data source that indicates the start and end times of a
+    validity window, where the dimension value is valid for any time within that range.
+    """
+
+    is_start: bool
+    is_end: bool
+
+
+class DimensionTypeParams(Protocol):
+    """Dimension type params add context to some types of dimensions (like time)"""
+
+    is_primary: bool
+    time_granularity: TimeGranularity
+    validity_params: Optional[DimensionValidityParams]
+
+
+class Dimension(Protocol):
```

Comment:
I do wonder if these instead of being implemented on an object (and specified as part of the protocol), if they should be defined in as functions that work on top of the implementing protocol. I know we generally don't like grab bags (hi @tlento), but maybe we should have general functions like the following implemented in DSI 
```python

from dbt_semantic_interfaces.protocols.dimension import Dimension
from dbt_semantic_interfaces.references import DimensionReference, TimeDimensionReference
from dbt_semantic_interfaces.type_enums.dimension_type import DimensionType

def is_primary_time(dim: Dimension) -> bool:  # noqa: D
    if dim.type == DimensionType.TIME and dim.type_params is not None:
        return dim.type_params.is_primary

    return False

def reference(dim: Dimension) -> DimensionReference:  # noqa: D
    return DimensionReference(element_name=dim.name)

def time_dimension_reference(dim: Dimension) -> TimeDimensionReference:  # noqa: D
    assert dim.type == DimensionType.TIME, f"Got type as {dim.type} instead of {DimensionType.TIME}"
    return TimeDimensionReference(element_name=dim.name)

def validity_params(dim: Dimension) -> Optional[DimensionValidityParams]:
    """Returns the DimensionValidityParams property, if it exists.

    This is to avoid repeatedly checking that type params is not None before doing anything with ValidityParams
    """
    if dim.type_params:
        return dim.type_params.validity_params

    return None
 ```
 
 I don't think it makes sense for these functions to be part of the Protocol spec. We could require these functions as part of the protocol, but then we require everyone to implement them. That is rather annoying since these really should always be implemented the same. We could define `@abstractmethod` for the functions on the protocol. However the implementing objects would only get these if they directly inherited the protocol, which we are avoiding. I think we're thus stuck with the example I outlined above 😬 Thoughts @tlento & @plypaul?

Reply:
Did we have an answer to this? I honestly can't remember, it's been quite a week and it's only Wednesday.

If not that's fine, we can address later. I agree these specific methods aren't really useful on the base interface, so we can push them into MetricFlow if we want, or provide a convenience mixin for default implementations, or whatever. I remember something about restructuring the type parameters to make some of these standard parts of a TimeDimension object.

Reply:
I believe we settled on putting these inside of a `convenience mixin` after yesterdays call, but @QMalcolm can correct me if I'm wrong here.

Reply:
@tlento The solution we tentatively decided on was to implement them in mixins. I'm still of the mind that they aren't necessary on the protocols and that it'd be easier to just provide the functions that take in dims, but I think I'm in the minority on that.

Reply:
We do plan on restructuring the type params into distinct dimension type specific protocols. However, I'm pretty much putting this PR on hold until after the cut over

Reply:
Sounds good! After a second look I agree with you on these 4 not being necessary for the core interface defined in the protocol spec - they're convenience methods - but it's good for me to see our approach for the more general form of the problem as well. Thanks all!